Plutarch/Rational.hs

```diff
@@ -340,3 +314,23 @@ pproperFraction = phoistAcyclic $
   plam $ \x ->
     plet (ptruncate # x) $ \q ->
       pcon $ PPair q (x - Plutarch.Rational.pfromInteger # q)
+
+-- Helpers
+
+cmpHelper ::
+  forall (s :: S).
+  Term s ((PInteger :--> PInteger :--> PBool) :--> PRational :--> PRational :--> PBool)
+cmpHelper = phoistAcyclic $ plam $ \f l r ->
+  pmatch l $ \(PRational ln ld) ->
+    pmatch r $ \(PRational rn rd) ->
+      f # (pto rd * ln) # (rn * pto ld)
+
+-- Assumes d is positive
+preduce' :: forall (s :: S). Term s (PInteger :--> PInteger :--> PRational)
+preduce' = phoistAcyclic $ plam $ \n d' ->
+  plet d' $ \d ->
+    plet (pgcd # n # d) $ \r ->
+      pif
+        (d #<= 0)
+        (pcon $ PRational (pnegate # (pdiv # n # r)) $ punsafeDowncast (pnegate # (pdiv # d # r)))
+        (pcon $ PRational (pdiv # n # r) $ punsafeDowncast (pdiv # d # r))
```

Comment:
> Assumes d is positive

But it doesn't right? It branches and negates if non-positive

Reply:
Huh, I am obviously blind. In that case, I can wring out even more.